New design

avni510 opened the pull request.

avni510 committed:
Removal of dead validation code

Removed valid conditions in board and console interface since those are
now handled by the Validation class. Deleted everything from Game since
a new design will be started

avni510 committed:
Symbol Loop Class and Tests

Created a Symbol Loop class that handles the loop related to choosing a
valid symbol. The tests use doubles but could be cleaned up. Tried
working on the GameLoop Class and tests, but tests are still failing.
EndGame has not been written yet

avni510 committed:
End Game Class

Added an End Game class that handles the end messages of a game. Also
made changes to GameLoop and SymbolLoop so their only dependencies are
the core logic. Work in Progress for the Game class run method and tests

avni510 committed:
Game Class

Finished writing and testing the Game class run method. The executable
needs to be updated.

avni510 committed:
Working version of Tic Tac Toe

Exectuble file was updated to include the new classes (GameLoop,
SymbolLoop, EndGame). This is a working version of Tic Tac Toe from the
command line.

avni510 committed:
Game Loop Tests

Retested the GameLoop class after pairing with Jason. Only one double is
used. The class also includes tests for a loop to enter a correct board
position.

avni510 committed:
Game Class Tests

Tested the game loop within the game class. Symbol Loop and End Game
still need to be tested in the Game class.

avni510 committed:
Symbol Loop Tests

Rewrote and added tests to the Symbol Loop Class and tested the Game
class. Still need to write tests for EndGame and incorporate that in
Game

avni510 committed:
End Game Tests in Game Class

added tests related to End Game within the Game class

Renamed ConsoleInterface to ConsoleUI and symbol to marker

avni510 committed:
End Game Tests in Game Class

added tests related to End Game within the Game class

avni510 committed:
Renamed ConsoleInterface

renamed ConsoleInterface to ConsoleUI and interface to console_ui

avni510 committed:
Renamed symbol to marker

Renamed symbol to marker because symbol is used in Ruby

avni510 committed:
Refactored GameLoop Tests

Refactored GameLoop Tests, made them more organized and got rid of code
duplication. Also created a player turn message that displays to the
user who's turn it is and which symbol they are. Also refactored
ConsoleUI and EndGame by deleting duplicate code.

avni510 committed:
Refactored GameLoop Tests

Refactored GameLoop Tests, made them more organized and got rid of code
duplication. Also created a player turn message that displays to the
user who's turn it is and which symbol they are

avni510 committed:
Refactored ConsoleUI and EndGame

Deleted duplicate code in ConsoleUI and end_game_spec

avni510 committed:
Reduced Duplication

Renamed 'instructions' to 'execution' to make it more clear. Also
deleted some duplicate code. Added a new way to test the game loop, but
not sure which is the proper way

avni510 committed:
Reduced Duplication

Renamed 'instructions' to 'execution' to make it more clear. Also
deleted some duplicate code. Added a new way to test the game loop, but
not sure which is the proper way

avni510 committed:
Restructed Game Loop tests

Reorganized GameLoop tests by game won, tied, or neither. Also renamed
.execution to .execute. Renamed GameLoop to GameTurn and GameTurn now only returns integers instead of integers and strings. Commented out parts of the test that
don't effect if the game_turn_spec.rb tests will pass or not.

avni510 committed:
deleted commented out code

avni510 committed:
Merge branch 'new_design' of github.com:avni510/tic_tac_toe into new_design

lisahamm commented on lib/marker_loop.rb:
```diff
@@ -0,0 +1,37 @@
+    end
+
+    def marker_assignment(p1_marker, game_marker1, game_marker2)
+      if p1_marker == game_marker1
```
Can you use the ternary operator here? I think you should be able to accomplish the goal of this method in 1 line as opposed to 5 lines if you do

avni510 replied:
I wasn't sure how to use a ternary operator here without losing the functionality of the until loop. I tried to use one with a begin loop but it ended up being the same number lines as the until loop. I was able to use a ternary operator for the proceeding method #marker_assignment

lisahamm commented on spec/game_turn_spec.rb:
```diff
@@ -0,0 +1,163 @@
+  require "game_turn"
+  require "board"
+  require "validation"
+  require "eval_game"
```
You can use a spec_helper file to require the classes/modules you need and then you will only need to require the spec_helper file in each of your test files as opposed to having a list of requires in your tests. This way, the file names are kept in only one place as opposed to being scattered throughout the tests. So if you change the name of a class/module, you will only need to update the  name in one place in order for your tests to still require everything properly.

lisahamm commented on spec/game_turn_spec.rb:
```diff
@@ -0,0 +1,163 @@
+      expect(game_turn).to receive(:valid_move_loop).and_return(move)
+    end
+    
+    
```
Extra line here can be removed

lisahamm commented on lib/game.rb:
```diff
@@ -1,69 +1,22 @@
-        @console_int.puts_space
+    def run 
+      marker1, marker2 =  @marker_loop.marker_selection
+      until @tied_or_player_won
```
@tied_or_player_won is really checking if the game is over. I would suggest not removing #game_over? from your EvalGame class and using the method here. This will be more explicit and simple to understand for anyone else who comes into your codebase to review or make changes.

lisahamm commented on lib/game.rb:
```diff
@@ -1,69 +1,22 @@
+    def run 
+      marker1, marker2 =  @marker_loop.marker_selection
+      until @tied_or_player_won
+        @tied_or_player_won = @game_turn.execute(1, marker1)
```
If you use the EvalGame's #game_over? to check the status, you will lose the logic that determines the game result in this class. To me, that seems like something that would be within the scope of EvalGame's responsibilities.

lisahamm commented on lib/game.rb:
```diff
@@ -1,69 +1,22 @@
+      until @tied_or_player_won
+        @tied_or_player_won = @game_turn.execute(1, marker1)
+        break if @tied_or_player_won
+        @tied_or_player_won = @game_turn.execute(2, marker2)
```
It also seems like you could DRY up this method a bit. You execute the game turn method twice here and also check if the game is over a couple of times. I wonder if you could reduce some of the duplication by keeping track of the player turn within this class and then you won't have to have the hard-coded player numbers/magic numbers within #run

lisahamm commented on lib/game_turn.rb:
```diff
@@ -0,0 +1,34 @@
+        @console_ui.display_board(@board.nine_space_array)
+        0
+      end
+    end
```
I don't think I would be expecting GameTurn#execute to return the game result. I would suggest removing the game result logic from this method and adding it into your EvalGame class since its responsibility is to evaluate the game. You can then evaluate the game result outside of this class

lisahamm commented on lib/game_turn.rb:
```diff
@@ -0,0 +1,34 @@
+    
+    private
+
+    def valid_move_loop(move)
```
I like the idea of this method a lot

lisahamm commented on lib/game_turn.rb:
```diff
@@ -0,0 +1,34 @@
+    private
+
+    def valid_move_loop(move)
+      until @validation.board_conditions?(@board.nine_space_array, move)
```
Can you think of a better name for the `@validation.board_conditions?` method? Without looking at it's implementation, it's not completely clear as to what the method is actually doing by the name

lisahamm commented on lib/eval_game.rb:
```diff
@@ -2,10 +2,6 @@ module TicTacToe
 
   class EvalGame
```
We usually use nouns for class names and verbs for method names. Maybe change this name to something like GameEval or GameEvalution

avni510 committed:
attempt at fixing errors in game_turn_spec.rb lines 147

avni510 committed:
ternary operator used for #marker_assignment, attempt for using a ternary operator in #marker_selection_loop commented out and does not work"

avni510 committed:
added spec_helper.rb

avni510 committed:
Moved player won or tied responsiblity

Moved player won or tied responsiblity from #execute method in GameTurn
class to #player_won_or_tied in EvalGame. #execute method now only
returns an array. Refactoring still needs to be done in the
eval_game_spec.rb, game_turn_spec.rb, and game.rb

avni510 committed:
Renamed #board_conditions? and #marker_conditions?

renamed these two methods to make it more clear.

avni510 committed:
renamed EvalGame to GameEval

avni510 committed:
refactored game.rb

refactored the #run method and found some areas where eval_game had not
been changed to game_eval. Also renamed #finished_game to
game_over_messages

avni510 committed:
refactored game_eval.rb

made #tied? and #won? methods private and changed tests accordingly

avni510 committed:
renamed EndGame to GameCompletion

avni510 committed:
deleted commented out code

also fixed tests related to validation_spec.rb

avni510 committed:
Created hash for GameTurn params

Implemented functionality so GameTurn can take in parameters in any
order. GameTurn now takes in a hash with parameters in the hash.

lisahamm commented on lib/game.rb:
```diff
@@ -1,69 +1,27 @@
+      marker1, marker2 =  @marker_loop.marker_selection
+      while true
+        break if game_over_after_player_turn?(1, marker1)
+        break if game_over_after_player_turn?(2, marker2)
```
Lines 15-16 have some duplication. Can you think of a way to DRY it up?

lisahamm commented on lib/game_turn.rb:
```diff
@@ -0,0 +1,27 @@
+  class GameTurn
+
+    def initialize(args)
+      @board = args[:board]
```
explore Hash#fetch

lisahamm commented on lib/marker_loop.rb:
```diff
@@ -0,0 +1,33 @@
+      @console_ui.select_marker_message
+      p1_marker = marker_selection_loop
+      p1_marker.upcase!
+      p2_marker = marker_assignment(p1_marker, "X", "O")
```
Can you set a constant or variable to the marker strings?

lisahamm commented on spec/spec_helper.rb:
```diff
@@ -0,0 +1,10 @@
+  require "marker_loop"
+  require "validation"
+  require "wrapper_io"
+
```
You can accomplish this by just requiring the TTT module

lisahamm commented on lib/game_completion.rb:
```diff
@@ -0,0 +1,19 @@
+    def game_over_messages(game_result, board_array)
+      @console_ui.display_board(board_array)
+      @console_ui.puts_space
+      if game_result == 1 || game_result == 2
```
Using integers 0, 1, and 2 to keep track of the game result leads to some confusion for anyone who comes in to read this code.

avni510 committed:
Created a Player class

* player oridinal numbers are more clear
* Player class keeps track of the current player and marker
* Player class switches between players
* reduced duplication in #run in game.rb

avni510 committed:
TTT constants for 'X' and 'O'

* created constants for the markers of TTT game
* allows for easier readability
* required tic_tac_toe.rb in spec_helper
* reduced duplication in #execute method

avni510 committed:
Incorporated usage of hash#fetch

* GameTurn does not require the board argument for initialization
* if the board argument doesn't exist it will create an instance of
Board

avni510 committed:
refactored Player class and fixed bug in game_eval.rb

avni510 committed:
Refactored Player Class

* each player in the game becomes a Player object
* #toggle_player is moved to game.rb

lisahamm commented on lib/game_completion.rb:
```diff
@@ -0,0 +1,19 @@
+    def game_over_messages(game_result, board_array)
+      @console_ui.display_board(board_array)
+      @console_ui.puts_space
+      if game_result == "player_one" || game_result == "player_two"
```
Since line #11 is checking if the game has been won, why not just ask the GameEval class instead of re-implementing that logic here?

lisahamm commented on lib/game_completion.rb:
```diff
@@ -0,0 +1,19 @@
+      @console_ui.puts_space
+      if game_result == "player_one" || game_result == "player_two"
+        @console_ui.won_message(game_result)
+      elsif game_result == "tied"
```
Same as above, the logic in line #13 already exists in the game eval class

lisahamm commented on lib/game_completion.rb:
```diff
@@ -0,0 +1,19 @@
+      @console_ui = console_ui
+    end
+
+    def game_over_messages(game_result, board_array)
```
What do you think about making use of your GameEval object inside of this method? You've already implemented #won? and #tie? and you could also add logic to return the winning player's mark. This should give you the benefit of not needing to use string values to indicate the game result, as well as eliminating some duplication of logic

avni510 committed:
Extracted game_result logic

* removed game_result logic from Game to GameCompletion
* removed duplication and hardcoded string values from game_result
